OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+
+    // Jacobian Matrices
+    SimTK::Mat22 df_dy;
+    SimTK::Mat22 df_dydot;
+
+
+    //-------Convert projFiberLength & Velocity to fiberLength & Velocity------/
+    double cosPenn;
+    double dcosPenn_dprojFibLen;
+    double fiberLengthNorm;
+    double dfiberLength_dprojFibLen;
+    if (pennAtOptFiberLength<0.01)  {
+        // If pennation is zero, we can't do this because volume is zero,
+        //      and fiberLength ~ projFiberLength
+        cosPenn=1.0;
+        fiberLengthNorm=projFibLenNorm;
+        dfiberLength_dprojFibLen=1;
+        dcosPenn_dprojFibLen=0;
+    }
+    else {
+        double b=sin(pennAtOptFiberLength);
+        fiberLengthNorm=sqrt(pow(projFibLenNorm,2) + pow(b,2));
+        cosPenn = projFibLenNorm/fiberLengthNorm;
+        dfiberLength_dprojFibLen=cosPenn;
+        dcosPenn_dprojFibLen=pow(b,2) / pow(fiberLengthNorm,3);
+
+    }
+
+    // Compute fiberVelocity and its derivatives wrt projFibLen and projFibVel
+    double fiberVelocityNorm = projFibVelNorm*cosPenn;
+    double dfiberVelocityNorm_dprojFibVelNorm = cosPenn;
+    double dfiberVelocityNorm_dprojFibLenNorm =
+            projFibVelNorm * dcosPenn_dprojFibLen;
+
+
+    //---F1 is the normalized isometric force-length relationship at maximum
+    //                                               activation--------------//
+    double fiberExp = (fiberLengthNorm - 1.0) / fl_width;   // [dimensionless]
+    double F1 = exp(-pow(fiberExp, 2));        // Gaussian force-length curve
+
+    double dF1_dfiberLengthNorm = 0;
+    if (returnJacobians) {
+        dF1_dfiberLengthNorm = -2.0 * fiberExp * F1 / fl_width;
+        double dF1_dprojFibLenNorm =
+                dF1_dfiberLengthNorm * dfiberLength_dprojFibLen;
+    }
+
+
+
+    //-------- F2 is the dimensionless force-velocity relationship -------//
+    double F2;
+    double dF2_dfiberVelocityNorm;
+    double dF2_dactiv;
+    double dF2_dprojFibVelNorm;
+    double dF2_dprojFibLenNorm;
+    double df_dmuscleLength;
+
+    // Chow/Darling Vel-Activation Relationship //TODO:  Add full reference
+    //double lambda = 0.5025 + 0.5341*activ;
+    //double  dlambda_da = 0.5341;
+    double lambda = 1;   //Turn it off for now as it seems to cause an issue
+                        // with negative force large concentric vel
+    double dlambda_da =0;
+
+    if (fiberVelocityNorm < 0) {
+        //Hill's equation for concentric contraction
+        // F2 = (V_{max} + V_{fiber}) / (V_{max} - V_{fiber}/a_{Hill})
+
+        double hillDenom = (lambda*vMaxNorm - fiberVelocityNorm/fv_AHill);
+        F2 = (lambda*vMaxNorm + fiberVelocityNorm) / hillDenom;
+
+        if (returnJacobians) {
+            dF2_dfiberVelocityNorm  = (1.0 + F2 / fv_AHill) / hillDenom;
+            dF2_dactiv = - dlambda_da * vMaxNorm * fiberVelocityNorm *
+                    (1.0 + 1.0/fv_AHill) / pow(hillDenom,2);
+        }
+    }
+    else {
+        //Katz Curve for eccentric contraction
```

Comment:
I will add one.  But FYI it Hill for Shortening and Katz for Lengthening:  Katz 1939 J Physio
http://onlinelibrary.wiley.com/doi/10.1113/jphysiol.1939.sp003756/epdf